Framework/Algorithms/src/EstimateDivergence.cpp

```diff
@@ -0,0 +1,125 @@
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string EstimateDivergence::category() const {
+  return "Diffraction\\Utility";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string EstimateDivergence::summary() const {
+  return "Estimate the divergence of each detector pixel";
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void EstimateDivergence::init() {
+  declareProperty(Kernel::make_unique<WorkspaceProperty<API::MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input),
+                  "Workspace to have divergence calculated from");
+
+  auto positiveParameter =
+      boost::make_shared<Kernel::BoundedValidator<double>>();
+  positiveParameter->setLower(0.);
+  positiveParameter->setLowerExclusive(false); // zero is allowed
+
+  declareProperty("alpha", 0., positiveParameter,
+                  "Vertical divergence parameter");
+  declareProperty("beta0", 0., positiveParameter,
+                  "Horizontal divergence parameter");
+  declareProperty("beta1", 0., positiveParameter,
+                  "Other horizontal divergence parameter");
+
+  declareProperty(
+      Kernel::make_unique<WorkspaceProperty<API::MatrixWorkspace>>(
+          "OutputWorkspace", "", Direction::Output),
+      "Workspace containing the divergence of each detector/spectrum");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void EstimateDivergence::exec() {
+  // read input parameters
+  API::MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
+  const double alpha = getProperty("alpha");
+  const double beta0 = getProperty("beta0");
+  const double beta1 = getProperty("beta1");
+
+  // derived terms
+  const double horizontal = alpha * alpha;
+  const double vertical_numerator = 4. * (beta0 * beta0 + beta1 * beta1);
+
+  // create output workspaces
+  API::MatrixWorkspace_sptr divergenceWS =
+      DataObjects::create<DataObjects::Workspace2D>(*inputWS,
+                                                    HistogramData::Points(1));
+
+  // do the math
+  const auto &spectrumInfo = inputWS->spectrumInfo();
+  const auto &detectorInfo = inputWS->detectorInfo();
+  const auto samplepos = spectrumInfo.samplePosition();
+  size_t numspec = inputWS->getNumberHistograms();
+  double solidangletotal = 0.;
+  for (size_t i = 0; i < numspec; ++i) {
+    // angle
+    const double twotheta =
+        spectrumInfo.isMonitor(i) ? 0.0 : spectrumInfo.twoTheta(i);
+    const double sintwotheta = sin(twotheta);
+    // vertical term
+    const double vertical = vertical_numerator / (sintwotheta * sintwotheta);
+
+    // solid angle
+    double solidangle = 0.0;
+    for (const auto detID : inputWS->getSpectrum(i).getDetectorIDs()) {
```

Comment:
Using detector IDs from the spectrum has a couple of issues:
- the detector might not exist, and you will get an exception
- does not work for workspaces with scanning instruments
- it is horribly slow to use `Detector`

You can try this instead:

```cpp
for (const auto &index : spectrumInfo.spectrumDefinition(i))
  // No scanning support for solidAngle currently, use only first component of index, ignore time index
  solidangle += componentInfo.solidAngle(index.first, samplepos);
```
Support for solid angle was added only recently by Owen, I think it is not yet available on the `DetectorInfo` API since there is no scanning support yet, but going via `ComponentInfo` is just as good (and much much faster than using `Detector`), if you do not require scanning.